source/common/http/conn_manager_impl.cc

```diff
@@ -1844,13 +1850,30 @@ void ConnectionManagerImpl::ActiveStream::refreshAccessLogFlushTimer() {
 }
 
 void ConnectionManagerImpl::ActiveStream::clearRouteCache() {
+  if (cached_route_blocked_) {
+    return;
+  }
+
+  if (hasCachedRoute()) {
+    // The configuration of the route may be referenced by some filters.
+    // Cache the route to avoid it's destroyed before the stream is destroyed.
+    cleared_cached_routes_.emplace_back(std::move(cached_route_.value()));
```

Comment:
OOC, if the filters had a shared_ptr to the route, would this still be necessary?

Reply:
If the filters cached the shared_ptr to the route, then this is not necessary. However, there is no any guard for that.
Most filter think all the route config will be cached by the active stream and will not destoryed until the request is completed.

Reply:
The `traversePerFilterConfig` or `mostSpecificPerFilterConfig` will not return shared_ptr but raw pointer or raw reference.

I think `header_to_metadata`, `header_mutation`, etc will store raw pointer to the route level filter config in the decoding phase and use it in the encoding phase. And I think there should be lots third party filters will do this like that.

It's easy to update the filters in our repo, but I am afraid this will breakdown lots third party filters. And this problem is hard to be found in the compile time.

Reply:
OK I see. In this particular case I would argue that we should change the type name to force a compile error and then have a release note about how to fix it. I'm not sure how large of a change that would be though. Perhaps by just changing the name of the type alias it would be enough?

Usually I would not argue for such a thing, but it seems not great to add this type of extra complexity when we have a way to force people to take notice and fix their code. WDYT?

Reply:
Do you mean change the basic type name to notify the third party developers that their filters are broken.

And then to tell them that they should keep a copy of route (or some others way) in the filter if they want to keep route level filter config's pointer in the filter by a release note?

Reply:
I'm saying to change the name of the fewest number of things that will cause all users of storing a route to have a compile break. We can fix all filters/extensions in the public repo, and then have a release note for other users to store a strong pointer. I think this is worth it in this case vs. carrying the additional lifetime hacks in the HCM. cc @alyssawilk for a second opinion on this.

Reply:
Sounds ok to me. But I think the things need to change wouldn't be 'few'.

we can update the return type of 'perFilterConfig' and the method name to get route.
Then almost all third party filters that try to store the route level config or route will get compile error.

Reply:
Yeah I'm not sure what would be the "easiest" thing to do here, but I do think this is the best path forward. Do you mind prototyping it to see how bad it is?

Reply:
I can have a try in a separated pr first to see how much code we need to change.